contracts/BatchAuction.sol

```diff
@@ -59,24 +65,45 @@ contract BatchAuction is Ownable {
 
         address tokenAddress = tokenIdToAddressMap[tokenIndex];
         require(tokenAddress != address(0), "Requested token is not registered");
-
         require(
             ERC20(tokenAddress).transferFrom(msg.sender, address(this), amount), 
             "Unsuccessful transfer"
         );
-        
-        if (depositIndex != block.number / 20) {
-            depositIndex = block.number / 20;
-            emit IncrementDepositIndex(depositIndex);
-        }
 
+        uint depositSlot = this.depositIndex();
         uint16 accountId = publicKeyToAccountMap[msg.sender];
         bytes32 nextDepositHash = sha256(
-            abi.encodePacked(depositHashes[depositIndex].shaHash, accountId, tokenIndex, amount)
+            abi.encodePacked(depositHashes[depositSlot].shaHash, accountId, tokenIndex, amount)
         );
 
-        depositHashes[depositIndex] = DepositState({shaHash: nextDepositHash, applied: false});
+        depositHashes[depositSlot] = DepositState({shaHash: nextDepositHash, applied: false});
+        emit Deposit(accountId, tokenIndex, amount, depositSlot);
+    }
+
+    function depositIndex() public view returns (uint) {
+        return block.number / 20;
+    }
+
+    function stateIndex() public view returns (uint) {
+        return stateRoots.length - 1;
+    }
 
-        emit Deposit(accountId, tokenIndex, amount, depositIndex);
+    function applyDeposits(
+        uint slot,
+        bytes32 _currDepositHash,
+        bytes32 _currStateRoot,
+        bytes32 _newStateRoot
+    )
+        public onlyOwner()
+    {   
+        require(slot < this.depositIndex(), "Deposit slot must exist and be inactive");
+        require(depositHashes[slot].applied == false, "Deposits already processed");
+        require(depositHashes[slot].shaHash != 0, "Deposit slot is empty");
+        require(depositHashes[slot].shaHash == _currDepositHash, "Incorrect Deposit Hash");
+        require(stateRoots[this.stateIndex()] == _currStateRoot, "Incorrect State Root");
+
+        stateRoots.push(_newStateRoot);        
+        depositHashes[slot].applied = true;
+        emit StateTransistion("applyDeposits", _currStateRoot, _newStateRoot);
```

Comment:
Nit: we will have to pay gas for the “applyDeposit” String that is emitted with each event. 
Therefore I think it’s cheaper to have a separate event name for each type of transition.
If there is a reason for having the same as should at least make the type an enum (so that it’s encoded as an int instead of a string)

Reply:
Yes cool. I wasn't aware of this.